File: kotlin/Checkout/src/Customer.kt
/**
 * This class represents a customer at the checkout line.
 *
 * It holds number of items in the cart for the customer,
 * and a unique id to identify the location of the customer.
 *
 * Initialize a customer instance with given properties,
 * and calculate the time required to serve the customer.
 *
 * @param id a unique customer id
 * @param cartSize the number of items in the cart
 */
class Customer(private val id: Int, val cartSize: Int) {
    /**
     * Time required to serve the customer.
     *
     * @return time to serve in seconds.
     */
    var timeToServe = 45 + 5 * cartSize; private set

    /**
     * Check if the customer is finished processing all the items in the cart.
     * On every call, it will reduce 1 second from total time
     * and check if the customer used all it's time.
     *
     * @return if the customer has done processing all the cart items.
     */
    val isDone get() = --timeToServe == 0

    /**
     * Display customer id, number of items and time required to serve.
     *
     * @return String return customer info.
     */
    override fun toString() = String.format(
        "[Customer%3d: Item:%3d, Time:%4ds]", id, cartSize, timeToServe
    )
}

File: flutter/checkout_queue/lib/customer.dart
/// A customer with total number of items in cart and an unique identifier.
class Customer {
  Customer({required this.id, required this.cartSize});

  final int id;
  final int cartSize;

  // Calculate and store time required to serve the customer.
  late int _timeToServe = 45 + 5 * cartSize;

  // Time required to serve the customer.
  int get timeToServe => _timeToServe;

  // Check weather all cart items has been processed or not.
  // On every request, reduce 1 second from total remaining time
  // and check if the customer used all it's time.
  bool get isDone => --_timeToServe == 0;

  // Dump customer data as formated string.
  @override
  String toString() => '[Customer $id: Items: $cartSize, Time: $timeToServe]';
}
